c2rust-refactor/src/scripting/ast_visitor.rs

```diff
@@ -0,0 +1,655 @@
+use rlua::prelude::{LuaContext, LuaError, LuaFunction, LuaResult, LuaTable};
+use syntax::ast::{
+    Arg, BindingMode, Block, Crate, Expr, ExprKind, FunctionRetTy, FnDecl, ImplItem,
+    ImplItemKind, Item, ItemKind, Local, Mod, Mutability::*, NodeId, Pat, PatKind,
+    Stmt, StmtKind,
+};
+use syntax::source_map::symbol::Symbol;
+use syntax::ptr::P;
+
+use crate::ast_manip::fn_edit::{FnKind, FnLike};
+use crate::scripting::{IntoLuaAst, TransformCtxt, utils::iter_to_lua_array};
+
+macro_rules! call_lua_visitor_method {
+    ($obj: expr , $method: ident ($($params: expr),*)) => {
+        let opt_visit_method: Option<LuaFunction> = $obj.get(stringify!($method))?;
+
+        if let Some(visit_method) = opt_visit_method {
+            let proceed = visit_method.call::<_, bool>(($obj.clone(), $($params.clone()),*))?;
+
+            if !proceed {
+                return Ok(());
+            }
+        }
+    };
+}
+
+pub(crate) struct LuaAstVisitor<'a, 'lua, 'tctx: 'a> {
+    ctx: &'a TransformCtxt<'a, 'tctx>,
+    lua_ctx: LuaContext<'lua>,
+    visitor_obj: LuaTable<'lua>
+}
+
+impl<'a, 'lua, 'tctx> LuaAstVisitor<'a, 'lua, 'tctx> {
+    pub fn new(visitor_obj: LuaTable<'lua>, ctx: &'a TransformCtxt<'a, 'tctx>, lua_ctx: LuaContext<'lua>) -> Self {
+        LuaAstVisitor {
+            ctx,
+            lua_ctx,
+            visitor_obj,
+        }
+    }
+
+    pub fn visit_crate(&self, krate: &mut Crate) -> LuaResult<()> {
+        let lua_krate = krate.clone().into_lua_ast(self.ctx, self.lua_ctx)?;
+
+        call_lua_visitor_method!(self.visitor_obj,visit_crate(lua_krate));
+
+        self.visit_module(lua_krate.get("module")?)?;
+        self.finish()?;
+
+        krate.merge_lua_ast(lua_krate)?;
```

Comment:
Is it correct that `merge_lua_ast` is only called when visiting the full crate, and not on any child nodes? That seems odd to me. What if I only visit a single function? I think the recursion should be removed from `merge_lua_ast` and put into the (Mut)Visitor functions.

Reply:
> Is it correct that merge_lua_ast is only called when visiting the full crate, and not on any child nodes? That seems odd to me. What if I only visit a single function?

Interesting point, I suppose I set it up such that `visit_crate` is the main entry point. Do you think it makes sense to leave merging to the caller of `visit_crate` (and friends) then?

> I think the recursion should be removed from merge_lua_ast and put into the (Mut)Visitor functions.

One of the big problems I found was that the `Mut`/`Visitor` traits don't return results, so our error handling gets completely borked. The lua code returns a ton of results and so need to be unwrapped or saved to self which seemed a really unwieldy.

Reply:
It would be nice to allow visiting sub-crate AST nodes and their children.

AFAICT you're not actually using the MutVisitor trait? I meant your `LuaAstVisitor` methods.